static/js/stream_data.js

```diff
@@ -500,6 +500,13 @@ export function is_notifications_stream_muted() {
     return is_muted(page_params.realm_notifications_stream_id);
 }
 
+export function can_subscribe_to(sub) {
+    // If stream is public then any user can subscribe. If stream is private then only
+    // subscribed users can unsubscribe.
+    // Guest users can't subscribe themselves to any stream.
+    return sub.subscribed || (!page_params.is_guest && !sub.invite_only);
+}
```

Comment:
I think `can_subscribe_to` isn't perfect for naming this function, because `can_subscribe_to` should be false if you're already subscribed -- whereas we want this function to return to so that we display an "Unsubscribe" button in these contexts.

Maybe `can_toggle_subscription` would be accurate?